tests/unit/models/gpflow/test_models.py

```diff
@@ -502,6 +503,82 @@ def test_gaussian_process_regression_optimize(
     assert internal_model.training_loss(**args) < loss
 
 
+@random_seed
+@pytest.mark.parametrize("after_model_optimize", [True, False])
+@pytest.mark.parametrize("after_model_update", [True, False])
+def test_gaussian_process_cached_predictions_correct(
+    after_model_optimize: bool,
+    after_model_update: bool,
+    gpflow_interface_factory: ModelFactoryType,
+) -> None:
+    x = np.linspace(0, 5, 10).reshape((-1, 1))
+    y = fnc_2sin_x_over_3(x)
+    data = x, y
+    dataset = Dataset(*data)
+    model, _ = gpflow_interface_factory(
+        x,
+        y,
+        optimizer=Optimizer(
+            gpflow.optimizers.Scipy(), minimize_args={"options": {"max_iter": 5}}, compile=True
+        ),
+    )
+
+    if isinstance(model, VariationalGaussianProcess):
+        pytest.skip("Cached predictions are not yet implemented for the VGP models.")
+
+    if after_model_optimize:
+        model.optimize(dataset)
+
+    if after_model_update:
+        new_x = np.linspace(0, 5, 3).reshape((-1, 1))
+        new_y = fnc_2sin_x_over_3(new_x)
+        new_dataset = Dataset(new_x, new_y)
+        model.update(new_dataset)
+        model.optimize(new_dataset)
+
+    x_predict = np.linspace(0, 5, 2).reshape((-1, 1))
+
+    # get cached predictions
+    cached_fmean, cached_fvar = model.predict(x_predict)
+    cached_joint_mean, cached_joint_var = model.predict_joint(x_predict)
+    cached_ymean, cached_yvar = model.predict_y(x_predict)
+
+    # get reference (slow) predictions from underlying model
+    reference_fmean, reference_fvar = model.model.predict_f(x_predict)
+    reference_joint_mean, reference_joint_var = model.model.predict_f(x_predict, full_cov=True)
+    reference_ymean, reference_yvar = model.model.predict_y(x_predict)
+
+    npt.assert_allclose(cached_fmean, reference_fmean)
+    npt.assert_allclose(cached_ymean, reference_ymean)
+    npt.assert_allclose(cached_joint_mean, reference_joint_mean)
+    npt.assert_allclose(cached_fvar, reference_fvar, atol=1e-5)
+    npt.assert_allclose(cached_yvar, reference_yvar, atol=1e-5)
+    npt.assert_allclose(cached_joint_var, reference_joint_var, atol=1e-5)
+    npt.assert_allclose(cached_yvar - model.get_observation_noise(), cached_fvar, atol=5e-5)
+
+
+@random_seed
+def test_gaussian_process_cached_predictions_faster(
+    gpflow_interface_factory: ModelFactoryType,
+) -> None:
+    x = np.linspace(0, 10, 10).reshape((-1, 1))
+    y = fnc_2sin_x_over_3(x)
+    model, _ = gpflow_interface_factory(x, y)
+
+    if isinstance(model, VariationalGaussianProcess):
+        pytest.skip("Cached predictions are not yet implemented for the VGP models.")
+
+    x_predict = np.linspace(0, 5, 2).reshape((-1, 1))
+    t_0 = time()
+    [model.predict(x_predict) for _ in range(10)]  # make sequential predictions
+    time_with_cache = time() - t_0
+    t_0 = time()
+    [model.model.predict_f(x_predict) for _ in range(10)]
```

Comment:
That doesn't look right. Would be much nicer if Trieste user could have a choice of using or not using posterior object. Of course unless it is 100% certain users will never want the latter.

This seems like a job for a [decorator pattern](https://refactoring.guru/design-patterns/decorator/python/example).

Reply:
I think we would always want to use them if they exist! Free  speeeed

Reply:
if there is really no downside to this, I also dont see a reason why not to make it a default